adjustments for readonly filesystems

Johann150 opened the pull request:
When running the driveboard app on something like a Raspberry Pi, from an SD card, it is desirable to use a read only filesystem to prolong the life of the SD card. These changes are mostly with Linux in mind.

- On a read-only filesystem, `mask`-ing or `unmask`-ing with systemd is not possible because you would have to write the symlinks to the files.
- Temporary directories on Linux are usually handled by an in-memory filesystem so are not affected by the read-only attribute of the root filesystem.

Johann150 committed:
use temporary directory for jobs

Some systems use a read-only file system, for example when running off the
SD card of a Raspberry Pi. In this case, writing to the home directory is
not possible. Instead, a temporary directory is created and used for
job file uploads.

Johann150 committed:
continue when unable to turn off hibernation

On read only file systems, trying to disable hibernation will fail because
no symlinks can be created. Instead of completely failing, print a warning
message and continue the program.

scottshambaugh commented:
Thanks @Johann150! I'm going to have to test this one back at home, my worry just looking at it is that there's value in having a default location for the config file that doesn't get stored in a temp directory. Does a config with modifications persist over power cycles?

scottshambaugh commented:
The hibernation part looks good though.

scottshambaugh commented:
Ok just tested, and it does look like the tempdir changes every time the program is run. Do you have another way you could implement a flag for read-only filesystems, and default to a consistent config directory otherwise?

Johann150 committed:
Revert "use temporary directory for jobs"

There are actual configuration files in the `confdir` which should
not be put in a temporary directory. There needs to be a separate
approach where the jobs are stored in a different directory from the
actual configuration file.

This reverts commit 47441656aea8ef01eaf467363b19de7dabae523c.

Johann150 commented:
Oh I wasn't aware that there was an actual configuration file, I only saw that the job files are stored in the same directory, my bad.

That configuration should probably not be done in a temporary directory, but also the jobs should have a separate or at least separately configurable directory. I'll have a look.

Johann150 committed:
use temporary directory for jobs

Separates the configuration directory from the queue directory, to be able to
use a temporary directory for the queue files.

Using a temporary directory for queue files is a good idea because some devices
may be using a read-only file system, e.g. when using a Raspberry Pi with an
SD card. In that case it will not be possible to write to locations in the home
directory or similar.

Writing to the configuration directory is not necessary during operation, but
writing to the queue directory is definitely necessary, so it makes sense to
separate the two.

Johann150 commented:
Okay I think it should be fixed now. I can also rebase the branch to remove the erroneous commit(s) if you prefer.

scottshambaugh commented:
Looks good to me now! Assumed you’ve tested locally, probably won’t be able to do so myself quickly. 

Thank you! Feel free to get in touch with anything to do with the project, I’m not sure how many of us there are using it right now.
